bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/PerChannelBookieClient.java

```diff
@@ -560,8 +562,7 @@ void addEntry(final long ledgerId, byte[] masterKey, final long entryId, ByteBuf
                     .setOperation(OperationType.ADD_ENTRY)
                     .setTxnId(txnId);
 
-            byte[] toSendArray = new byte[toSend.readableBytes()];
-            toSend.getBytes(toSend.readerIndex(), toSendArray);
+            byte[] toSendArray = toSend.toArray();
```

Comment:
We still have a buffer copy into a newly allocated array, and this is happening per bookie, WQ times.

Reply:
Yes, this is because protobuf cannot take a direct memory buffer with the payload.

Reply:
Why can't we have the change done at pendingAddOp level and avoid 3 copies? These are immutable buffers, no need to have WQ copies.

Reply:
we can replace ByteString.copyFrom(toSendArray) with UnsafeByteOperations.unsafeWrap(..). IIRC  copyFrom() may end up actually doing copy.
 
Also please look into using combo of UnsafeByteOperations.unsafeWrap-ing each ByteBuf separately with ByteString.concat on top. This should skip extra allocations (I haven't tried that yet)

Reply:
I agree with the above comments, though I'd rather leave this outside the scope of this PR which is a 1-1 replacement for `DoubleByteBuf`. The `byte[]` copies are not changing with this change.

Reply:
ok, LGTM then.